modules/drivers/athena/src/metabase/driver/athena.clj

```diff
@@ -305,68 +304,86 @@
-   Such as prepare statement"
-  [database query]
-  (log/infof "Running Athena query : '%s'..." query)
-  (try
-    (jdbc/query (sql-jdbc.conn/db->pooled-connection-spec database) (str/replace query ";" " ") {:raw? true})
-    (catch Exception e
-      (log/error (u/format-color 'red "Failed to execute query: %s %s" query (.getMessage e))))))
-
-(defn- describe-database->clj
-  "Workaround for wrong getColumnCount response by the driver (huh?)"
-  [column-metadata]
-  {:name (str/trim (:col_name column-metadata))
-   :type (str/trim (:data_type column-metadata))})
-
-(defn- remove-invalid-columns
-  "Returns a transducer."
-  []
-  (comp (remove #(= (:col_name %) ""))
-        (remove #(= (:col_name %) nil))
-        (remove #(= (:data_type %) nil))
-        (remove #(str/starts-with? (:col_name %) "#")) ; remove comment
-        (distinct)                                     ; driver can return twice the partitioning fields
-        (map describe-database->clj)))
-
-(defn- describe-table-fields-with-nested-fields [database schema table-name]
+(defn- reducible-fields-from-metadata
+  [^DatabaseMetaData metadata ^String db-name-or-nil ^String schema ^String table-name]
+  (reify clojure.lang.IReduceInit
+    (reduce [_this rf init]
+      (with-open [rset (.getColumns metadata db-name-or-nil schema table-name nil)]
+        (loop [acc init]
+          (if-not (.next rset)
+            acc
+            (recur (rf acc (merge {:name          (str/trim (.getString rset "COLUMN_NAME"))
+                                   :database-type (str/trim (.getString rset "TYPE_NAME"))}
+                                  (when-let [remarks (.getString rset "REMARKS")]
+                                    (when (not (str/blank? remarks))
+                                      {:field-comment remarks})))))))))))
+
+(defn- add-database-position-xform []
+  (map-indexed (fn [i field]
+                 (assoc field :database-position i))))
+
+(defn- fields-from-metadata-xform [driver]
+  (comp (map (fn [{:keys [database-type], :as field}]
+               (assoc field :base-type (database-type->base-type-or-warn driver database-type))))
+        (add-database-position-xform)))
+
+(defn- reducible-fields-from-query
+  [database schema table-name]
+  (reify clojure.lang.IReduceInit
+    (reduce [_this rf init]
+      (with-open [conn (jdbc/get-connection (sql-jdbc.conn/db->pooled-connection-spec database))
+                  stmt (doto (.createStatement conn)
+                         (.execute (format "DESCRIBE `%s`.`%s`;" schema table-name)))
+                  rset (.getResultSet stmt)]
+        (loop [acc init]
+          (if-not (.next rset)
+            acc
+            (recur (rf acc {:name          (str/trim (.getString rset "COL_NAME"))
+                            :database-type (str/trim (.getString rset "DATA_TYPE"))}))))))))
+
+(defn- fields-from-query-xform [driver]
```

Comment:
Athena gets metadata in two different ways depending on whether there are nested columns or not, the logic is a little different between the two ways but untangling that and figuring out how we could consolidate them more was outside of the scope of this PR I think. I did improve things a little by having the two versions use consistent keys but it could still use more cleanup in the future.